File: benchmark/synduce/list_to_tree/search_v2.f
Inductive Tree = empty Unit | node {Int, Tree, Tree};
Inductive List = nil Unit | cons {Int, List};

cat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {h, t} -> cons {h, f t y}
  | nil _ -> y
  end
);

repr =
  let dec = fix (
    \f: List -> Tree -> List. \res: List. \t: Tree.
    match t with
      empty _ -> res
    | node {w, l, r} -> f (cons {w, f res r}) l
    end
  ) in fix (
    \f: Tree -> List. \t: Tree.
    match t with
      empty _ -> nil unit
    | node {w, l, r} -> dec (cons {w, f r}) l
    end
  );

target = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    empty _ -> empty unit
  | node {w, l, r} -> node {w, f l, f r}
  end
);

@Input x: Int;

spec = fix (
  \f: List -> Bool. \xs: List.
  match xs with
    nil _ -> false
  | cons {h, t} -> if == h x then true else f t
  end
);

main = \xs: Tree. spec (repr (target xs));

File: ml-benchmark/synduce/list_to_tree/search_v2.ml
type tree = Empty | Node of int * tree * tree
type listt = Nil | Cons of int * listt

let rec cat x y =
  match x with
  | Cons (h, t) -> Cons (h, cat t y)
  | Nil -> y

let rec dec res t =
  match t with
  | Empty -> res
  | Node (w, l, r) ->
      let tmp = dec res r in
      let res2 = Cons (w, tmp) in
      dec res2 l

let rec repr t =
  match t with
  | Empty -> Nil
  | Node (w, l, r) ->
      let rr = repr r in
      let res = Cons (w, rr) in
      dec res l

let rec target t =
  match t with
  | Empty -> Empty
  | Node (w, l, r) ->
      let l2 = target l in
      let r2 = target r in
      Node (w, l2, r2)

let x = 0

let rec spec xs =
  match xs with
  | Nil -> false
  | Cons (h, t) ->
      if h = x then true else spec t

let program xs = spec (repr (target xs))